### DISCUSSION TITLE
AWS Lambda with S3 deployment

### DISCUSSION THREAD

**Discussion by AbraamNader** [directus/directus#6259] [Archive]
Dears,

Directus is awesome, so I am trying to make a proof of concept so our company would adopt using it for some upcoming projects.

I am trying to deploy Directus to AWS Lambda, and use AWS S3 for uploads folder, however I have some questions:

Does Directus require write access to the local file system for uploads to work? if so, I think AWS lambda wouldn't work.

Also since trying from localhost to upload to S3 using the following .env variables:
```
STORAGE_LOCATIONS="AWS"
STORAGE_AWS_DRIVER="s3"
STORAGE_AWS_ROOT="/uploads"
STORAGE_AWS_KEY="**KEY***"
STORAGE_AWS_SECRET="***SECRET***"
STORAGE_AWS_REGION="eu-west-1"
STORAGE_AWS_BUCKET="mybucketname.com"
```

and removed the local storage location env variables.

I get these errors:

- On uploading an image without a folder (on the root directory)
```{
  "errors": [
    {
      "message": "insert into [directus_files] ([filename_download], [folder], [id], [storage], [title], [type], [uploaded_by]) values (@p0, @p1, @p2, @p3, @p4, @p5, @p6) - Conversion failed when converting from a character string to uniqueidentifier.",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR"
      }
    }
  ]
}
```

- On uploading an image within a folder, or using a field uploader, the file is uploaded to S3 successfully, but without the other sizes,
so when getting the thumbnail or sizes on the admin, the following error is thrown in the browser console:

```
{"errors":[{"message":"E_UNKNOWN: An unknown error happened with the file uploads/457cabc3-01ce-432c-aaae-63a4326c2b70__fit_cover,height_300,key_system-medium-cover,width_300.png.\n\nError code: Forbidden\nOriginal stack:\nForbidden: null\n    at Request.extractError (\\node_modules\\aws-sdk\\lib\\services\\s3.js:698:35)\n    at Request.callListeners (\\node_modules\\aws-sdk\\lib\\sequential_executor.js:106:20)\n    at Request.emit (\\node_modules\\aws-sdk\\lib\\sequential_executor.js:78:10)\n    at Request.emit (\\node_modules\\aws-sdk\\lib\\request.js:688:14)\n    at Request.transition (\\node_modules\\aws-sdk\\lib\\request.js:22:10)\n    at AcceptorStateMachine.runTo (\\node_modules\\aws-sdk\\lib\\state_machine.js:14:12)\n    at \\node_modules\\aws-sdk\\lib\\state_machine.js:26:10\n    at Request.<anonymous> (\\node_modules\\aws-sdk\\lib\\request.js:38:9)\n    at Request.<anonymous> (\\node_modules\\aws-sdk\\lib\\request.js:690:12)\n    at Request.callListeners (\\node_modules\\aws-sdk\\lib\\sequential_executor.js:116:18)","extensions":{"code":"INTERNAL_SERVER_ERROR"}}]}
```

- On deleting the uploaded image in the admin, I get the following error:

```
{"errors":[{"message":"E_UNKNOWN: An unknown error happened with the file uploads/457cabc3-01ce-432c-aaae-63a4326c2b70.\n\nError code: AccessDenied\nOriginal stack:\nAccessDenied: Access Denied\n    at Request.extractError (\\node_modules\\aws-sdk\\lib\\services\\s3.js:718:35)\n    at Request.callListeners (\node_modules\\aws-sdk\\lib\\sequential_executor.js:106:20)\n    at Request.emit (\\node_modules\\aws-sdk\\lib\\sequential_executor.js:78:10)\n    at Request.emit (\\node_modules\\aws-sdk\\lib\\request.js:688:14)\n    at Request.transition (\\node_modules\\aws-sdk\\lib\\request.js:22:10)\n    at AcceptorStateMachine.runTo (\\node_modules\\aws-sdk\\lib\\state_machine.js:14:12)\n    at \\node_modules\\aws-sdk\\lib\\state_machine.js:26:10\n    at Request.<anonymous> (\\node_modules\\aws-sdk\\lib\\request.js:38:9)\n    at Request.<anonymous> (\\node_modules\\aws-sdk\\lib\\request.js:690:12)\n    at Request.callListeners (\\node_modules\\aws-sdk\\lib\\sequential_executor.js:116:18)","extensions":{"code":"INTERNAL_SERVER_ERROR"}}]}
```

and the file entry is removed from the database however it remains on the S3 bucket.

I wish the docs showed the required permissions for S3 to work with Directus:
IAM User Policy
Bucket policy
Bucket settings
Cross-origin resource sharing (CORS)
ACL
...

for now I just gave it all the unrestricted options just to make it work, like "s3:*", "*" etc.

Would appreciate any help with this.

Version: 9.0 RC 75 (upgraded from RC 74, using the bootstrap command)
Database: SQL Server
Cache: Redis (auto purge enabled)
Installation: Manual install on NodeJS 14 localhost (Windows) 
Browser: Edge 91

**Comment by rijkvanzanten** [comment]
> Does Directus require write access to the local file system for uploads to work? if so, I think AWS lambda wouldn't work.

No, only if you're using the local file storage adapter does directus need local file access

> On uploading an image without a folder (on the root directory)

This sounds like a bug in Directus

The other two issues do sound like a permissions issue for S3 access though. Both that `Forbidden` and `Access denied` are returned by S3. The first is thrown when Directus tries to read if the file exists on the bucket, and the second when it tries generating / saving the thumbnail 🤔

**Reply by AbraamNader** [reply]
This answers my first question thank you, also thank you for the quick reply.

Aside from the possible root folder bug, the other two access issues are weird, how can Directus create the file on S3, but has no read/delete access and can't create the thumbnails?

If you can provide any S3 permission settings that are required by Directus, I would much appreciate it.
though, I don't know what more can I do, since I used wildcards for most of the settings.

**Reply by rijkvanzanten** [reply]
That's a very good question, to which I don't have the answer. The only thing I know is that those issues you're seeing are thrown by S3, which can only lead me to conclude a permissions issue in there..

**Reply by AbraamNader** [reply]
I finally managed to get it working!
Thank you Rijk for pointing me to the right direction 👍 
it was the IAM user attached policy was , so this worked:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "s3:*",
            "Resource": [
                "arn:aws:s3:::mybucketnameexample.com/*",
                "arn:aws:s3:::mybucketnameexample.com"
            ]
        }
    ]
}
```

I wish there was more specific permissions that we can use other than `s3:*`. I could just go the trial and error way; add and remove permissions and see what works.

My next challenge is to host it on AWS lambda.

**Reply by rijkvanzanten** [reply]
> I wish there was more specific permissions that we can use other than s3:*

I'm sure there is! 

I'm no expert in AWS's permission rules etc.. That being said, we only use a specific sub-set of the aws-sdk methods for file management (see everything prefixed with `this.$driver` here: https://github.com/directus/directus/blob/main/packages/drive-s3/src/AmazonWebServicesS3Storage.ts) so I'm sure there's a more dedicated sub-set of permissions we could extract from that.

> My next challenge is to host it on AWS lambda.

Lemme know how it goes! I've been making small tweaks left and right to make sure connections aren't opened until needed etc to (hopefully) start supporting serverless environments better. It's definitely an area i'm interested in, so lemme know if and when you run into something funky!

**Reply by AbraamNader** [reply]
from the storage service I can see these:
s3:CopyObject
s3:DeleteObject
s3:HeadObject
s3:GetObject
I will check it out.

however, I'm stuck in the AWS Lambda issue, 
I created a serverless framework template here:
https://github.com/AbraamNader/directus-serverless-template

and as you can see I am still clueless since I am not a nodejs developer, 
the only experience I have with nodejs is from a frontend point of view, mainly angular with SSR using serverless framework.

I am trying to pass the express app that Directus creates to the serverless express plugin handler, but I think this is not how it works, I will have to study more to understand what I am supposed to do.

If it doesn't work out, plan B would be to use beanstalk instead, but I wish it works with lambda, since its cheaper, so my company would be more likely to invest in using Directus.

**Reply by mouhannad-sh** [reply]
Hi @AbraamNader  👋🏻 
I'm attempting to do a similar thing to what you were trying to do here.
if you don't mind sharing, did it work out in the end ? did you face any major blockers ?

**Reply by AbraamNader** [reply]
Hi @mouhannad-sh,

if you mean uploading to s3, then yes everything worked great.

but if you mean the lambda deployment, then no it did not work, also I don't think AWS lambda is right use case for this, 
even when using Redis caching, the start up time for the lambda function might be an issue.
In the end I just used EC2.

May be AWS App Runner will be a more suitable and cost efficient solution,
however I am waiting for AWS to fix this major issue:
https://github.com/aws/apprunner-roadmap/issues/1

**Reply by rijkvanzanten** [reply]
I would also suggest looking into ECS to run Directus. Way easier to manage than a "raw" EC2 instance 🙂
